### PR TITLE
Add IsString instance for ShortText newtypes

### LINKED ISSUE #4224: Add IsString UnqualComponentName instance (and other no-more-strings)
What it says on the tin.

### PR DESCRIPTION
**Pull request by phadej**
Resolve #4224

### CODE REVIEW

File: Cabal/Distribution/ModuleName.hs [hunk 1]
```diff
@@ -76,12 +76,12 @@ simple str = ModuleName (stlFromStrings [str])
...
-fromString :: String -> ModuleName
-fromString string = fromComponents (split string)
-  where
-    split cs = case break (=='.') cs of
-      (chunk,[])     -> chunk : []
-      (chunk,_:rest) -> chunk : split rest
+instance IsString ModuleName where
+    fromString string = fromComponents (split string)
+      where
+        split cs = case break (=='.') cs of
```
Comment (23Skidoo) [inline Cabal/Distribution/ModuleName.hs]: May be worth copying `splitOn` from `Data.List.Split` to `D.S.Utils`.

Comment (phadej) [inline Cabal/Distribution/ModuleName.hs]: Let's do it later. This commit just moves as little code as possible

Comment (23Skidoo) [inline Cabal/Distribution/ModuleName.hs]: Sure.

**Comment by ezyang** [comment]
OK!

**Comment by 23Skidoo** [comment]
Travis failure is genuine:

```
Distribution/Utils/ShortText.hs:20:1: Warning:
    The import of ‘Data.String’ is redundant
      except perhaps to import instances from ‘Data.String’
    To import instances alone, use: import Data.String()
<no location info>: 
Failing due to -Werror.
```

File: Cabal/Distribution/ModuleName.hs (hunk 1 above)
Comment (phadej) [inline Cabal/Distribution/ModuleName.hs]: Now i remember, `splitOn` is quite general reusing other combinators in `split`. Also this is less general `splitOn' :: a -> [a] -> [[a]]`, where general one splits on `[a]`.

**Comment by phadej** [comment]
new try

File: Cabal/Distribution/ModuleName.hs (hunk 1 above)
Comment (23Skidoo) [inline Cabal/Distribution/ModuleName.hs]: You're right, this is more like `Data.List.wordsBy`.

**Comment by phadej** [comment]
Ah. cannnot bump `optparse-applicative` because of https://github.com/well-typed/hackage-security/pull/179. Will wait for it.

**Pull request merged**